### ISSUE TITLE
add_cluster_node fails with 9.04 

### ISSUE THREAD

**Issue by sygilber** [IBM-Security/ibmsecurity#46]
It is working fine with ISAM 9.0.3 and ISAM 9.0.3.1

Ansible trace I am getting.

TASK [add_cluster_node : Add xxxxyyyyp.iad.ca.inet to cluster] ******************************************************************************************************************************************************************************
fatal: [xxxxyyyyp.iad.ca.inet]: FAILED! => {"changed": false, "failed": true, "log": "[2017-12-02 13:31:57,180] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.user.applianceuser] [__init__():9] Creating a user\n[2017-12-02 13:31:57,180] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.user.user] [__init__():8] Creating a user\n[2017-12-02 13:31:57,180] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.isamappliance] [__init__():13] Creating an ISAMAppliance\n[2017-12-02 13:31:57,180] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [__init__():15] Creating an IBMAppliance\n[2017-12-02 13:31:57,181] [PID:31939 TID:140203096975168] [INFO] [ibmsecurity.appliance.ibmappliance] [_log_desc():30] *** Retrieving firmware ***\n[2017-12-02 13:31:57,181] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_warnings():92] Checking for minimum version: None.\n[2017-12-02 13:31:57,181] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_warnings():108] Checking for one of required modules: None.\n[2017-12-02 13:31:57,181] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_warnings():125] Warnings: []\n[2017-12-02 13:31:57,181] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_invoke_request():301] Headers are: {'Content-type': 'application/json', 'Accept': 'application/json'}\n[2017-12-02 13:31:57,181] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_invoke_request():306] Input Data: {}\n[2017-12-02 13:31:57,181] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_suppress_ssl_warning():35] Suppressing SSL Warnings.\n[2017-12-02 13:31:57,181] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_url():24] Issuing request to: https://xxxxyyyyp.iad.ca.inet:443/firmware_settings\n[2017-12-02 13:31:57,183] [PID:31939 TID:140203096975168] [DEBUG] [urllib3.connectionpool] [_new_conn():824] Starting new HTTPS connection (1): xxxxyyyyp.iad.ca.inet\n[2017-12-02 13:31:57,763] [PID:31939 TID:140203096975168] [DEBUG] [urllib3.connectionpool] [_make_request():396] https://xxxxyyyyp.iad.ca.inet:443\"GET /firmware_settings HTTP/1.1\" 200 505\n[2017-12-02 13:31:57,765] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_response():63] Status Code: 200\n[2017-12-02 13:31:57,766] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_response():65] Text: [{\"partition\":\"1\",\"last_boot\":\"Never\",\"name\":\"isam_9.0.4.0_b4_20171102-2205_beta\",\"active\":false,\"comment\":\"\",\"id\":1,\"firmware_version\":\"IBM Security Access Manager 9.0.4.0_b4\",\"install_date\":\"1512146776\",\"backup_date\":null,\"install_type\":\"ISO\"}, {\"partition\":\"2 (Active)\",\"last_boot\":\"1512232344\",\"name\":\"isam_9.0.4.0_b4_20171102-2205_beta\",\"active\":true,\"comment\":\"\",\"id\":2,\"firmware_version\":\"IBM Security Access Manager 9.0.4.0_b4\",\"install_date\":\"1512146795\",\"backup_date\":null,\"install_type\":\"ISO\"}]\n[2017-12-02 13:31:57,766] [PID:31939 TID:140203096975168] [INFO] [ibmsecurity.appliance.ibmappliance] [_log_desc():30] *** Get Setup Complete Settings ***\n[2017-12-02 13:31:57,766] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_warnings():92] Checking for minimum version: None.\n[2017-12-02 13:31:57,766] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_warnings():108] Checking for one of required modules: None.\n[2017-12-02 13:31:57,766] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_warnings():125] Warnings: []\n[2017-12-02 13:31:57,766] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_invoke_request():301] Headers are: {'Content-type': 'application/json', 'Accept': 'application/json'}\n[2017-12-02 13:31:57,766] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurit
y.appliance.ibmappliance] [_invoke_request():306] Input Data: {}\n[2017-12-02 13:31:57,766] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_suppress_ssl_warning():35] Suppressing SSL Warnings.\n[2017-12-02 13:31:57,766] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_url():24] Issuing request to: https://xxxxyyyyp.iad.ca.inet:443/setup_complete\n[2017-12-02 13:31:57,767] [PID:31939 TID:140203096975168] [DEBUG] [urllib3.connectionpool] [_new_conn():824] Starting new HTTPS connection (1): xxxxyyyyp.iad.ca.inet\n[2017-12-02 13:31:58,254] [PID:31939 TID:140203096975168] [DEBUG] [urllib3.connectionpool] [_make_request():396] https://xxxxyyyyp.iad.ca.inet:443\"GET /setup_complete HTTP/1.1\" 200 21\n[2017-12-02 13:31:58,257] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_response():63] Status Code: 200\n[2017-12-02 13:31:58,257] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_response():65] Text: { \"configured\": true}\n[2017-12-02 13:31:58,257] [PID:31939 TID:140203096975168] [INFO] [ibmsecurity.appliance.ibmappliance] [_log_desc():30] *** Retrieving activations ***\n[2017-12-02 13:31:58,258] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_warnings():92] Checking for minimum version: None.\n[2017-12-02 13:31:58,258] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_warnings():108] Checking for one of required modules: None.\n[2017-12-02 13:31:58,258] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_warnings():125] Warnings: []\n[2017-12-02 13:31:58,258] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_invoke_request():301] Headers are: {'Content-type': 'application/json', 'Accept': 'application/json'}\n[2017-12-02 13:31:58,258] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_invoke_request():306] Input Data: {}\n[2017-12-02 13:31:58,258] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_suppress_ssl_warning():35] Suppressing SSL Warnings.\n[2017-12-02 13:31:58,258] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_url():24] Issuing request to: https://xxxxyyyyp.iad.ca.inet:443/isam/capabilities/v1\n[2017-12-02 13:31:58,259] [PID:31939 TID:140203096975168] [DEBUG] [urllib3.connectionpool] [_new_conn():824] Starting new HTTPS connection (1): xxxxyyyyp.iad.ca.inet\n[2017-12-02 13:31:58,702] [PID:31939 TID:140203096975168] [DEBUG] [urllib3.connectionpool] [_make_request():396] https://xxxxyyyyp.iad.ca.inet:443\"GET /isam/capabilities/v1 HTTP/1.1\" 200 None\n[2017-12-02 13:31:58,705] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_response():63] Status Code: 200\n[2017-12-02 13:31:58,705] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_response():65] Text: [{\"name\":\"ISAM Base Appliance\",\"description\":\"IBM Security Access Manager Base Appliance\",\"id\":\"wga\",\"enabled\":\"True\"}, {\"name\":\"ISAM Advanced Access Control\",\"description\":\"IBM Security Access Manager Advanced Access Control\",\"id\":\"mga\",\"enabled\":\"True\"}, {\"name\":\"ISAM Federation\",\"description\":\"IBM Security Access Manager Federation\",\"id\":\"federation\",\"enabled\":\"True\"}]\n[2017-12-02 13:31:58,706] [PID:31939 TID:140203096975168] [INFO] [ibmsecurity.appliance.ibmappliance] [_log_desc():30] *** Retrieve the cluster identifier ***\n[2017-12-02 13:31:58,706] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_warnings():92] Checking for minimum version: None.\n[2017-12-02 13:31:58,706] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_warnings():108] Checking for one of required modules: None.\n[2017-12-02 13:31:58,706] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_warnings():12
5] Warnings: []\n[2017-12-02 13:31:58,706] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_invoke_request():301] Headers are: {'Content-type': 'application/json', 'Accept': 'application/json'}\n[2017-12-02 13:31:58,706] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_invoke_request():306] Input Data: {}\n[2017-12-02 13:31:58,706] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_suppress_ssl_warning():35] Suppressing SSL Warnings.\n[2017-12-02 13:31:58,706] [PID:31939 TID:140203096975168] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_url():24] Issuing request to: https://xxxxyyyyp.iad.ca.inet:443/isam/cluster/id/v2\n[2017-12-02 13:31:58,707] [PID:31939 TID:140203096975168] [DEBUG] [urllib3.connectionpool] [_new_conn():824] Starting new HTTPS connection (1): xxxxyyyyp.iad.ca.inet\n[2017-12-02 13:31:59,425] [PID:31939 TID:140203096975168] [DEBUG] [urllib3.connectionpool] [_make_request():396] https://xxxxyyyyp.iad.ca.inet:443\"GET /isam/cluster/id/v2 HTTP/1.1\" 404 0\n[2017-12-02 13:31:59,427] [PID:31939 TID:140203096975168] [ERROR] [ibmsecurity.appliance.ibmappliance] [_process_response():46]   Request failed: \n[2017-12-02 13:31:59,428] [PID:31939 TID:140203096975168] [ERROR] [ibmsecurity.appliance.ibmappliance] [_process_response():47]      status code: 404\n", "msg": "('HTTP Return code: 404', u'')", "name": "ibmsecurity.isam.base.cluster.node.add"}

**Comment by sygilber** [comment]
I just tested with ISAM 9.0.4 GA and the result in the same as the 9.0.4 beta. Apparently, it fails in the get_id() function (before even the add() is invoked) and it seems it does not line the return code:

The full traceback is:
  File "/tmp/ansible_SKDlVU/ansible_module_isam.py", line 110, in main
    ret_obj = eval(func_call)
  File "<string>", line 1, in <module>
  File "/home/someuser/.local/lib/python2.7/site-packages/ibmsecurity/isam/base/cluster/node.py", line 37, in add
    ret_obj = get_id(isamAppliance)
  File "/home/someuser/.local/lib/python2.7/site-packages/ibmsecurity/isam/base/cluster/node.py", line 20, in get_id
    "/isam/cluster/id/v2")
  File "/home/someuser/.local/lib/python2.7/site-packages/ibmsecurity/appliance/isamappliance.py", line 354, in invoke_get
    requires_version=requires_version, warnings=warnings)
  File "/home/someuser/.local/lib/python2.7/site-packages/ibmsecurity/appliance/isamappliance.py", line 323, in _invoke_request
    self._process_response(return_obj=return_obj, http_response=r, ignore_error=ignore_error)
  File "/home/someuser/.local/lib/python2.7/site-packages/ibmsecurity/appliance/isamappliance.py", line 51, in _process_response
    raise IBMError("HTTP Return code: {0}".format(http_response.status_code), http_response.text)

fatal: [somehost]: FAILED! => {
    "changed": false,
    "failed": true,
    "invocation": {
        "module_args": {
            "action": "ibmsecurity.isam.base.cluster.node.add",
            "appliance": "somehost",
            "force": false,
            "isamapi": {
                "restricted": true,
                "signature_file": "/home/someuser/isam_home/tmp/primary.sign"
            },
            "lmi_port": 443,
            "log": "DEBUG",
            "password": "*******",
            "username": "username"
        }
    },
    "log": "[2017-12-22 11:50:14,476] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.user.applianceuser] [__init__():9] Creating a user\n[2017-12-22 11:50:14,476] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.user.user] [__init__():8] Creating a user\n[2017-12-22 11:50:14,476] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.isamappliance] [__init__():13] Creating an ISAMAppliance\n[2017-12-22 11:50:14,476] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [__init__():15] Creating an IBMAppliance\n[2017-12-22 11:50:14,476] [PID:54765 TID:140227166996288] [INFO] [ibmsecurity.appliance.ibmappliance] [_log_desc():30] *** Retrieving firmware ***\n[2017-12-22 11:50:14,477] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_warnings():92] Checking for minimum version: None.\n[2017-12-22 11:50:14,477] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_warnings():108] Checking for one of required modules: None.\n[2017-12-22 11:50:14,477] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_warnings():125] Warnings: []\n[2017-12-22 11:50:14,477] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_invoke_request():301] Headers are: {'Content-type': 'application/json', 'Accept': 'application/json'}\n[2017-12-22 11:50:14,477] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_invoke_request():306] Input Data: {}\n[2017-12-22 11:50:14,477] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_suppress_ssl_warning():35] Suppressing SSL Warnings.\n[2017-12-22 11:50:14,477] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_url():24] Issuing request to: https://somehost:443/firmware_settings\n[2017-12-22 11:50:14,479] [PID:54765 TID:140227166996288] [DEBUG] [urllib3.connectionpool] [_new_conn():824] Starting new HTTPS connection (1): somehost\n[2017-12-22 11:50:15,002] [PID:54765 TID:140227166996288] [DEBUG] [urllib3.connectionpool] [_make_request():396] https://somehost:443 \"GET /firmware_settings HTTP/1.1\" 200 491\n[2017-12-22 11:50:15,006] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_response():63] Status Code: 200\n[2017-12-22 11:50:15,006] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_response():65] Text: [{\"partition\":\"1 (Active)\",\"last_boot\":\"1513724227\",\"name\":\"isam_9.0.4.0_20171201-2325_dev\",\"active\":true,\"comment\":\"\",\"id\":1,\"firmware_version\":\"IBM Security Access Manager 9.0.4.0\",\"install_date\":\"1512195276\",\"backup_date\":null,\"install_type\":\"ISO\"}, {\"partition\":\"2\",\"last_boot\":\"Never\",\"name\":\"isam_9.0.4.0_20171201-2325_dev\",\"active\":false,\"comment\":\"\",\"id\":2,\"firmware_version\":\"IBM Security Access Manager 9.0.4.0\",\"install_date\":\"1512195306\",\"backup_date\":null,\"install_type\":\"ISO\"}]\n[2017-12-22 11:50:15,006] [PID:54765 TID:140227166996288] [INFO] [ibmsecurity.appliance.ibmappliance] [_log_desc():30] *** Get Setup Complete Settings ***\n[2017-12-22 11:50:15,006] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_warnings():92] Checking for minimum version: None.\n[2017-12-22 11:50:15,007] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_warnings():108] Checking for one of required modules: None.\n[2017-12-22 11:50:15,007] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_warnings():125] Warnings: []\n[2017-12-22 11:50:15,007] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_invoke_request():301] Headers are: {'Content-type': 'application/json', 'Accept': 'application/json'}\n[2017-12-22 11:50:15,007] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_invoke_request():306] Input Data: {}\n[2017-12-22 11:50:15,007] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_suppress_ssl_warning():35] Suppressing SSL Warnings.\n[2017-12-22 11:50:15,007] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_url():24] Issuing request to: https://somehost:443/setup_complete\n[2017-12-22 11:50:15,008] [PID:54765 TID:140227166996288] [DEBUG] [urllib3.connectionpool] [_new_conn():824] Starting new HTTPS connection (1): somehost\n[2017-12-22 11:50:15,424] [PID:54765 TID:140227166996288] [DEBUG] [urllib3.connectionpool] [_make_request():396] https://somehost:443 \"GET /setup_complete HTTP/1.1\" 200 21\n[2017-12-22 11:50:15,427] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_response():63] Status Code: 200\n[2017-12-22 11:50:15,427] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_response():65] Text: { \"configured\": true}\n[2017-12-22 11:50:15,427] [PID:54765 TID:140227166996288] [INFO] [ibmsecurity.appliance.ibmappliance] [_log_desc():30] *** Retrieving activations ***\n[2017-12-22 11:50:15,427] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_warnings():92] Checking for minimum version: None.\n[2017-12-22 11:50:15,427] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_warnings():108] Checking for one of required modules: None.\n[2017-12-22 11:50:15,428] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_warnings():125] Warnings: []\n[2017-12-22 11:50:15,428] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_invoke_request():301] Headers are: {'Content-type': 'application/json', 'Accept': 'application/json'}\n[2017-12-22 11:50:15,428] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_invoke_request():306] Input Data: {}\n[2017-12-22 11:50:15,428] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_suppress_ssl_warning():35] Suppressing SSL Warnings.\n[2017-12-22 11:50:15,428] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_url():24] Issuing request to: https://somehost:443/isam/capabilities/v1\n[2017-12-22 11:50:15,429] [PID:54765 TID:140227166996288] [DEBUG] [urllib3.connectionpool] [_new_conn():824] Starting new HTTPS connection (1): somehost\n[2017-12-22 11:50:15,854] [PID:54765 TID:140227166996288] [DEBUG] [urllib3.connectionpool] [_make_request():396] https://somehost:443 \"GET /isam/capabilities/v1 HTTP/1.1\" 200 None\n[2017-12-22 11:50:15,857] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_response():63] Status Code: 200\n[2017-12-22 11:50:15,857] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_response():65] Text: [{\"name\":\"ISAM Base Appliance\",\"description\":\"IBM Security Access Manager Base Appliance\",\"id\":\"wga\",\"enabled\":\"True\"}]\n[2017-12-22 11:50:15,857] [PID:54765 TID:140227166996288] [INFO] [ibmsecurity.appliance.ibmappliance] [_log_desc():30] *** Retrieve the cluster identifier ***\n[2017-12-22 11:50:15,857] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_warnings():92] Checking for minimum version: None.\n[2017-12-22 11:50:15,858] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_warnings():108] Checking for one of required modules: None.\n[2017-12-22 11:50:15,858] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_process_warnings():125] Warnings: []\n[2017-12-22 11:50:15,858] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_invoke_request():301] Headers are: {'Content-type': 'application/json', 'Accept': 'application/json'}\n[2017-12-22 11:50:15,858] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_invoke_request():306] Input Data: {}\n[2017-12-22 11:50:15,858] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_suppress_ssl_warning():35] Suppressing SSL Warnings.\n[2017-12-22 11:50:15,858] [PID:54765 TID:140227166996288] [DEBUG] [ibmsecurity.appliance.ibmappliance] [_url():24] Issuing request to: https://somehost:443/isam/cluster/id/v2\n[2017-12-22 11:50:15,859] [PID:54765 TID:140227166996288] [DEBUG] [urllib3.connectionpool] [_new_conn():824] Starting new HTTPS connection (1): somehost\n[2017-12-22 11:50:16,367] [PID:54765 TID:140227166996288] [DEBUG] [urllib3.connectionpool] [_make_request():396] https://somehost:443 \"GET /isam/cluster/id/v2 HTTP/1.1\" 404 0\n[2017-12-22 11:50:16,369] [PID:54765 TID:140227166996288] [ERROR] [ibmsecurity.appliance.ibmappliance] [_process_response():46]   Request failed: \n[2017-12-22 11:50:16,370] [PID:54765 TID:140227166996288] [ERROR] [ibmsecurity.appliance.ibmappliance] [_process_response():47]      status code: 404\n",
    "msg": "('HTTP Return code: 404', u'')",
    "name": "ibmsecurity.isam.base.cluster.node.add"
}

This is using the latest ibmsecuruty code release this week.

I will try obtaining an HTTP trace.

**Comment by sygilber** [comment]
Here are 2 traces. Trace-ansible.log from ansible side with "-vvv" and trace-isam.log is isam appliance trace (with 'mesa' trace capture).

It is possible that starting with 9.04 that when the target appliance on which the call "GET /isam/cluster/id/v2" is submitted and that is not a cluster member yet, that it returns a "404" instead of an empty array ? Which would cause the python code to break.

[trace-ansible.log](https://github.com/IBM-Security/ibmsecurity/files/1583361/trace-ansible.log)
[trace-isam.log](https://github.com/IBM-Security/ibmsecurity/files/1583362/trace-isam.log)

Using the LMI to add the node to the cluster is working. By examining the trace, a call to "GET /isam/cluster/id/v2" is made and it also returns a 404 but the UX somehow seems to ignore that and continues.

So I am more and more convinced that 9.04 may something return a 404 rc to say the same thing as "no value" available. The python code will need therefore to ignore the 404 RC in this particular case and continue. This is not clean.

**Comment by sygilber** [comment]
I have perform locally a quick hack inside ibmsecurity.isam.base.cluster.node.py as follow to skip over the exception returned from the 404 and now it works.  

Of course, the clean solution should instead be to catch the IBMError raised and filter on rc=404. For any other returned HTTP RC, "re-throw" the error. Not sure how this is done in Python.

def add(isamAppliance, signature_file, restricted=False, check_mode=False, force=False):
    """
    Add a node to the cluster
    """
    id = None
    if force is False:
        try:
            ret_obj = get_id(isamAppliance)
            id = ret_obj['data']['value']
        except:
            pass
	
    if force is True or _check(isamAppliance, id) is False:
        if check_mode is True:
            return isamAppliance.create_return_object(changed=True)
        else:
            return isamAppliance.invoke_post_files(
                "Add a node to the cluster",
                "/isam/cluster/nodes/v1",
                [
                    {
                        'file_formfield': 'signature_file',
                        'filename': signature_file,
                        'mimetype': 'application/octet-stream'
                    }
                ],
                {'restricted': restricted}, json_response=False)

    return isamAppliance.create_return_object()

**Comment by ram-ibm** [comment]
@sygilber - the issue is documented:

```
404 Not found

 This response is returned when there is not an appropriate static IPv4 address configured on interface 1.1.
```

https://www.ibm.com/support/knowledgecenter/SSPREK_9.0.4/com.ibm.isam.doc/develop/rapi/Retrieve%20the%20cluster%20identifier.xml

**Comment by ram-ibm** [comment]
I am going to see if I can work the code to handle that situation more elegantly - after all it does so in lower versions.

**Comment by sygilber** [comment]
Resolved by commit 29b188a7487e32195aa27e7a8bbff6ccbad69904
